### DISCUSSION TITLE
QA Billing Service and Stripe

### DISCUSSION THREAD

**Discussion by anamarn** [twentyhq/twenty#8678] [Engineering vision]
**Why?**

This discussion has been created in order to create an internal document wich resumes how to test Billing Service. Because we use an external service (Stripe) and each error will be expensive it is necessary to create an extensive Manual test suit case. We syncronize our data with the stripe webhook, however we need to be careful when we write to our database and how we process the information. 

**Set Up for manual testing**

REMEMBER TO ALWAYS USE TEST MODE IN STRIPE

Stripe Webhook:

- Authenticate with your account in the stripe CLI
- Run the command: stripe listen --forward-to http://localhost:3000/billing/webhooks

Billing:

- Set IS_BILLING_ENABLED to true
- Add your BILLING_STRIPE_SECRET and BILLING_STRIPE_API_KEY
- Add your BILLING_STRIPE_BASE_PLAN_PRODUCT_ID (use the one in testMode > Base Plan)

On twenty Site:

- Buy a subscription (you can use the card 4242...42 with expiration date later in the future)

**Ideas of Situations**

- I use my account, I have 3 members in my workspace, I buy subscriptions for all of them
- I have a subscription and I invite 3 more people who accept the invitation
- I use my account , I use my second workspace, and I buy a subscription
- I buy a subscription and I cancel the payment, 
- I create a new account with only me in the workspace, I invite two people, Another user deletes me from the workspace

**Considerations**

_Entitlements:_

The data from the Entitlement table is updated based on the stripe webhook responses, and we use the customerActiveEntitlemet response to update the info on the table, however this event doesn't have the metadata containing the workspaceId. Because we cannot control at witch order the webhook send events, we force a server error if the entitlements are updated before the BillingSubscription Table. Stripe resends the event based on a exponential backoff (for more info see https://docs.stripe.com/webhooks#retries ) because we are in test mode Stripe retries three times over a few hours. So if the BillingEntitlement is not updated it is completely normal and it will be updated when stripe resends the event. Normally the customerActiveEntitlement Event only gives us the customerId, so we need to fetch the workspaceId associated from the BillingSubscription Table.

**Comment by BOHEUS** [comment]
@anamarn thanks a lot for this doc as it clears some of unknowns from [Lucas' comment](https://github.com/twentyhq/twenty/issues/6641#issuecomment-2421827072), I tried to test it on my own but couldn't progress further as Stripe requires creating account for company, not for person, I'll take care of it in the future after discussing how to test it properly without risking additional costs
